src/hotspot/share/runtime/cpuTimeCounters.hpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023 Google LLC. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+
+#ifndef SHARE_RUNTIME_CPUTIMECOUNTERS_HPP
+#define SHARE_RUNTIME_CPUTIMECOUNTERS_HPP
+
+
+#include "memory/iterator.hpp"
+#include "runtime/os.hpp"
+#include "runtime/perfData.hpp"
+#include "runtime/perfDataTypes.hpp"
+
+class CPUTimeGroups : public AllStatic {
+public:
+  enum class CPUTimeType {
+    gc_total,
+    gc_parallel_workers,
+    gc_conc_mark,
+    gc_conc_refine,
+    gc_service,
+    vm,
+    conc_dedup,
+    COUNT,
+  };
+
+  static const char* to_string(CPUTimeType val);
+  static bool is_gc_counter(CPUTimeType val);
+};
+
+class CPUTimeCounters: public CHeapObj<mtServiceability> {
+private:
+  // CPUTimeCounters is a singleton instance.
+  CPUTimeCounters();
+  NONCOPYABLE(CPUTimeCounters);
+
+  static CPUTimeCounters* _instance;
```

Comment:
I thought it is typically preferred to initialize a singleton object on the heap, rather than using many static variables. It is easier to control the initialization order and timing of an on-heap singleton object than statics.

Moreover, for this class, `initialize()` could also check `if (UsePerfData)`, and only create the singleton object under `UsePerfData`. This could save some memory when `UsePerfData` is false.

Reply:
I would say it depends on the use-case and here when switching to use static functions to use the instance it felt more like an all-static class. I agree that it would be nice to avoid the additional memory usage if `UsePerfData` is `false` so I'm ok with keeping the instance if we add that.

Reply:
Added the check in the initializer